Improved Kuadrant installation via Olm  and kustomize (#1075)
Kuadrant/kuadrant-operator: config/install/README.md

Add the following to your local directory.  For the purpose of this doc, we will use: `kuadrant/configure/`.

```yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
resources:
  - https://github.com/Kuadrant/kuadrant-operator//config/install/configure/standard?ref=v1.0.1 #change this version as needed (see https://github.com/Kuadrant/kuadrant-operator/releases)

```

Lets apply this to your cluster:

```bash

kubectl apply -k <kuadrant/configure>

```

### Verify kuadrant is installed and ready:

```bash
kubectl get kuadrant kuadrant -n kuadrant-system -o=wide

# NAME       STATUS   AGE
# kuadrant   Ready    109s

```

You should see the condition with type `Ready` with a message of `kuadrant is ready`. 

### Verify Istio is configured and ready:

```bash
kubectl get istio -n gateway-system

#sample output
# NAME      REVISIONS   READY   IN USE   ACTIVE REVISION   VERSION   AGE
# default   1           1       1        Healthy           v1.23.0   3d22h
```

At this point Kuadrant is installed and ready to be used as is Istio as the gateway provider. This means AuthPolicy and RateLimitPolicy can now be configured and used to protect any Gateways you create. 

> ## Configure a DNS provider credentials and a Certificate Issuer

Author: not needed. Is optional

> ## Configure a DNS provider credentials and a Certificate Issuer

Reviewer: who do we want the audience of this doc to be? what are we trying to show or teach people? 

My thinking is if we document different parts of Kuadrant as optional even tho they technically might be, its going to be a pain for us in the long run with keeping things up to date, making docs follow the same installation etc, are we just overcomplicating things?  

Should we just show how to install kuadrant as a whole and if they don't want to use a certain aspect then so be it. We could show in the architecture what components come out of the box and what are optional. 

side note: I think we need to have a convo as a team on how we want to portray the project in our docs as we are currently writing in all different styles and ways. I'm currently writing a doc (doception 😂 ) to suggest a standard and have a kick off point for conversation. 

This install guide is a bit of a double whammy because we also need a convo on what installation methods do we want to stand behind. we have many different types and different docs on how to do it and duplicate docs that go old and obsolete

This document will walk you through setting up the required configuration to install kaudrant using [kustomize](https://kustomize.io/) or a tool that leverages kustomize such as kubectl along with OLM. It will also go through more advanced configuration options to enable building up a resilient configuration. You can view the full configuration built here: [Full AWS Example](https://github.com/Kuadrant/kuadrant-operator/tree/main/config/install/full-example-aws).

1. [Basic Install](#basic-installation)

2. [Configure DNS and TLS integration](#configure-dns-and-tls-integration)

3. [External Redis for Rate Limit Counters](#use-an-external-redis)

4. [Limitador Resilient Configuration](#limitador-topologyconstraints-poddisruptionbudget-and-resource-limits)

5. [Authorino Resilient Configuration](#authorino-topologyconstraints-poddisruptionbudget-and-resource-limits)

4. [[OpenShift Specific] Setup Observability ](#set-up-observability-openshift-only)

## Prerequisites  
- OCP or K8s cluster and CLI available.
- OLM installed [operator lifecycle manager releases](https://github.com/operator-framework/operator-lifecycle-manager/releases)

> - (Optional) Gateway Provider Installed: By default this guide will install the [Sail Operator](https://github.com/istio-ecosystem/sail-operator) that will configure and install an Istio installation. Kuadrant is intended to work with [Istio](https://istio.io) or [Envoy Gateway](https://gateway.envoyproxy.io/) as a gateway provider before you can make use of Kuadrant one of these providers should be installed.  

Reviewer 2: ```suggestion
```
To avoid confusion, I'd remove this optional dependency because the guide shows how to install kuadrant with istio/sail. The 2 are entwined in the steps below, such that if the user already has a gateway provider installed that they want to use, there is extra work to decouple the istio/sail pieces.

Alternatively, I'm wondering if we should decouple those steps for the user anyways in this guide?
I think it's reasonable that a user could arrive at the Kuadrant installlation guide with or without a gateway provider already running in their cluster.

Author: Yeah but perhaps we can add that as an follow up. I think there is a lot of value in the guide as is for now?

- (Optional) cert-manager:
  - [cert-manager Operator for Red Hat OpenShift](https://docs.openshift.com/container-platform/4.16/security/cert_manager_operator/cert-manager-operator-install.html)
  - [installing cert-manager via OperatorHub](https://cert-manager.io/docs/installation/operator-lifecycle-manager/)

> - (Optional) Access to AWS, Azure or GCP with DNS services. 

Reviewer 2: For this guide, it's just AWS that's involved. (see comment below about marking that step as optional).
I'd remove mention of Azure and GCP here, or alternatively add info about how you'd use them instead of AWS in the relevant sections below.

Author: I will add a link to the examples for other providers

patches: # remove this subscription patch if you are installing a development version. It will then use the "preview" channel
  - patch: |-
      apiVersion: operators.coreos.com/v1alpha1
      kind: Subscription
      metadata:
        name: kuadrant
      spec:
        source: kuadrant-operator-catalog
        sourceNamespace: kuadrant-system
        name: kuadrant-operator
        channel: 'stable' #set to preview if not using a release (for example if using main)

```

And execute the following to apply it to a cluster:

```bash
# change the location depending on where you created the kustomization.yaml
kubectl apply -k $KUADRANT_DIR/install

```

#### Verify the operators are installed:

OLM should begin installing the dependencies for Kuadrant. To wait for them to be ready, run:

```bash
kubectl -n kuadrant-system wait --timeout=160s --for=condition=Available deployments --all
```

> Note: you may see ` no matching resources found ` if the deployments are not yet present.

Once OLM has finished installing the operators (this can take several minutes). You should see the following in the kuadrant-system namespace:

```bash
kubectl get deployments -n kuadrant-system

> ## Output

Reviewer 2: ```suggestion
## Output (kuadrant-console-plugin deployment only installed on OpenShift)
```

Kuadrant has 2 additional operand components that it manages: `Authorino` that provides data plane auth integration and `Limitador` that provides data plane rate limiting. To set these up lets add a new `kustomization.yaml` in a new sub directory. We will re-use this later for further configuration. We do this as a separate step as we want to have the operators installed first.

Add the following to your local directory.  For the purpose of this doc, we will use: `$KUADRANT_DIR/configure/kustomization.yaml`.

```bash
mkdir -p $KUADRANT_DIR/configure
touch $KUADRANT_DIR/configure/kustomization.yaml

```

Add the following to the new kustomization.yaml:

```yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
resources:
  - https://github.com/Kuadrant/kuadrant-operator//config/install/configure/standard?ref=v1.0.1 #change this version as needed (see https://github.com/Kuadrant/kuadrant-operator/releases)

```

Lets apply this to your cluster:

```bash

kubectl apply -k $KUADRANT_DIR/configure

```

### Verify kuadrant is installed and ready:

```bash
kubectl get kuadrant kuadrant -n kuadrant-system -o=wide

# NAME       STATUS   AGE
# kuadrant   Ready    109s

```

> You should see the condition with type `Ready` with a message of `kuadrant is ready`. 

Reviewer 2: ```suggestion
You should see the condition with type `Ready`. 
```
message not show in wide view

### Verify Istio is configured and ready:

```bash
kubectl get istio -n gateway-system

#sample output
# NAME      REVISIONS   READY   IN USE   ACTIVE REVISION   VERSION   AGE
# default   1           1       1        Healthy           v1.23.0   3d22h
```

At this point Kuadrant is installed and ready to be used as is Istio as the gateway provider. This means AuthPolicy and RateLimitPolicy can now be configured and used to protect any Gateways you create. 

> ## Configure DNS and TLS integration

Reviewer 2: ```suggestion
## (Optional) Configure DNS and TLS integration
```

To match the "(Optional) cert-manager" and "(Optional) Access to AWS" prereqs at the top, 
otherwise CRDs don't exist and no credentials.

First you will need to setup the required `.env` file specified in the kuztomization.yaml file in the same directory as your existing configure kustomization. Below is an example for AWS:

```bash
touch $KUADRANT_DIR/configure/aws-credentials.env

```
Add the following to your new file

```
AWS_ACCESS_KEY_ID=xxx
AWS_SECRET_ACCESS_KEY=xxx
AWS_REGION=eu-west-1

```

With this setup, lets update our configure kustomization to generate the needed secrets. We will also define a TLS ClusterIssuer (see below). The full `kustomization.yaml` file should look like:

```yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
resources:
  - https://github.com/Kuadrant/kuadrant-operator//config/install/configure/standard?ref=v1.0.1 #change this version as needed (see https://github.com/Kuadrant/kuadrant-operator/releases)
  - cluster-issuer.yaml #(comment if you dont want to use it. The issuer yaml is defined below). Ensure you name the file correctly.
  

generatorOptions:
  disableNameSuffixHash: true
  labels:
    app.kubernetes.io/part-of: kuadrant
    app.kubernetes.io/managed-by: kustomize

secretGenerator:
  - name: aws-provider-credentials

>     namespace: cert-manager # assumes cert-manager namespace exists.

Reviewer 2: ```suggestion
    namespace: cert-manager # change this to the namespace where cert-manager is running.
```
In my case, following olm install of cert-manager, it in the `operators` namespace

generatorOptions:
  disableNameSuffixHash: true
  labels:
    app.kubernetes.io/part-of: kuadrant
    app.kubernetes.io/managed-by: kustomize

secretGenerator:
  - name: aws-provider-credentials
    namespace: cert-manager # assumes cert-manager namespace exists.
    envs:
      - aws-credentials.env # notice this matches the .env file above. You will need to setup this file locally
    type: 'kuadrant.io/aws'
  - name: aws-provider-credentials
    namespace: gateway-system # this is the namespace where your gateway will be provisioned
    envs:
      - aws-credentials.env #notice this matches the .env file above. you need to set up this file locally first.
    type: 'kuadrant.io/aws'
  - name: redis-credentials
    namespace: kuadrant-system
    envs:
      - redis-credentials.env
    type: 'kuadrant.io/redis'

patches:
  - patch: |-
      apiVersion: limitador.kuadrant.io/v1alpha1
      kind: Limitador
      metadata:
        name: limitador
        namespace: kuadrant-system
      spec:
        storage:
          redis:
            configSecretRef:
              name: redis-credentials

```

Re-Apply the configuration to setup the new secret and configuration:

```bash

> kubectl apply -k $KUADRANT_DIR/configure/

Reviewer 2: Probably related to changes below https://github.com/Kuadrant/kuadrant-operator/pull/1075/files#diff-36411ee7f394bd28aaa769d3b33d193cde1e03f8f7bb3fd569b2efb1b4a509f0R4
would need to land that on the v1.0.1 branch

> kubectl apply -k $KUADRANT_DIR/configure/

Author: Yes I want to cherry pick these changes to the v1.0.1 branch. In the validation steps I mentioned you would have to use the ref=olm-installation for the resources .
